### ISSUE TITLE
Support executing wasm binaries

### ISSUE THREAD

**Issue by jjcollinge** [dapr/dapr#770]
## In what area(s)?
> /area runtime

## Describe the feature
Given the broad scope of environments Dapr aims to target, I believe it'll be important to support additional models beyond running as a side car container or process. I think, the vast majority of people don't _currently_ have the need to switch out components between environments and therefore could argue the performance overhead of this additional hop might outweight the benefit of using Dapr's abstracted APIs over native APIs. To address perf concerns, you/we may consider using something like [Wasmer](https://github.com/wasmerio/go-ext-wasm) or [Wagon](https://github.com/go-interpreter/wagon) to execute the user's app (compiled to wasm) directly in the runtime and then look to leverage some protocol layer hooks to try and avoid networking and [de]serialization for the local calls, such as some of the work discussed in [grpc-go](https://github.com/grpc/grpc-go/issues/906). I totally, appreciate this is a fairly immature set of technologies and is a premature optimization at this point but just wanted to see and track whether there was any thinking in this space.

**Comment by yaron2** [comment]
We've talked previously about compiling Dapr into a Go [wasm executable](https://github.com/golang/go/wiki/WebAssembly) and having it be consumable from wasm supported languages.

As you rightly pointed out, *currently*  the state of wasm (and server side WebAssembly in particular) is still not mature enough to allow us this kind of exploration.

Coupled with the fact switching out components between environments is not Dapr's main goal, it will be some time before we look into this again, but of course if users/customers say otherwise, that'll dictate the roadmap which currently doesn't include wasm. (personally, I'd like to see this happen).

Given the above, can we close the issue until it becomes relevant?

Thanks.

**Comment by jjcollinge** [comment]
Yeh, makes sense not to be a priority at this point in time. If I can find some spare time I may have a play with some of the experimental bits and see if I can get something working anyway.

**Comment by codefromthecrypt** [comment]
I figured I would bump a status on this as it seems the first request to dapr for WebAssembly (historical milestone!)

FYI wasi (GOOS=wasip1) will be available in Go 1.21 https://go-review.googlesource.com/c/go/+/479627

Here's an example of playing with it
```bash
go install golang.org/dl/gotip@latest
gotip download
brew install wazero
cat > main.go <<EOF
package main

func main() {
	println("hello")
}
EOF
GOARCH=wasm GOOS=wasip1 gotip build -o main.wasm main.go
wazero run main.wasm
```

The current implementation can only export the main function, so inputs and outputs are via args, env or stdio. The first place you can use this is an [output binding](https://github.com/dapr/components-contrib/issues/2274). Depending on what happens between now and 1.21, it might be usable for [http middleware](https://github.com/dapr/samples/tree/master/hello-wasm).

Meanwhile, TinyGo works for both of these use cases. Main thing is to install tinygo and use slightly different arguments when compiling (e.g. `tinygo build -o main.wasm -target=wasi main.go`).

As this area is fast moving in different directions, you can look at the #webassembly discord channel. Also, @daixiang0 @salaboy and I are doing a presentation at [KubeCon EU](https://kccnceu2023.sched.com/event/1HyW0/safe-dynamic-middleware-with-dapr-and-webassembly-mauricio-salatino-diagrid-adrian-cole-tetrate) next week (on the http middleware part)
